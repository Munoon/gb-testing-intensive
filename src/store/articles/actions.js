import * as types from "./types";

export const getArticles = (count) => ({
  type: types.GET_ARTICLES,
  payload: {
    count,
  },
});

export const getArticlesRequest = () => ({
  type: types.GET_ARTICLES_REQUEST,
});

export const getArticlesSuccess = (articles) => ({
  type: types.GET_ARTICLES_SUCCESS,
  payload: {
    articles,
  },
});

export const getArticlesFailure = (error) => ({
  type: types.GET_ARTICLES_FAILURE,
  payload: {
    error,
  },
});

export const toggleFavourite = (id) => ({
  type: types.TOGGLE_FAVOURITE,
  payload: {
    id
  }
})
