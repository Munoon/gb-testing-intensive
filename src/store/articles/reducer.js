import { REQUEST_STATUS } from "../../constants";
import * as types from './types';

const initialState = {
  articles: [],
  request: {
    status: REQUEST_STATUS.IDLE,
    error: null,
  },
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case types.GET_ARTICLES_REQUEST: return {
      ...state,
      request: {
        ...state.request,
        status: REQUEST_STATUS.LOADING
      }
    };
    case types.GET_ARTICLES_SUCCESS: return {
      articles: action.payload.articles,
      request: {
        error: null,
        status: REQUEST_STATUS.SUCCESS
      }
    };
    case types.GET_ARTICLES_FAILURE: return {
      ...state,
      request: {
        error: action.payload.error,
        status: REQUEST_STATUS.FAILURE,
      }
    };
    case types.TOGGLE_FAVOURITES: {
      const newArticles = [...state.articles];
      const index = newArticles.findIndex(({id}) => id === action.payload.id);

      if (!newArticles[index]) {
        return state
      }

      newArticles[index] = {
        ...newArticles[index],
        favourite: !newArticles[index].favourite
      }

      return {
        ...state,
        articles: newArticles
      }
    }
    default: return state;
  }
}