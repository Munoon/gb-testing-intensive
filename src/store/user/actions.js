import * as types from "./types";

export const setName = (name) => ({
  type: types.SET_NAME,
  payload: {
    name,
  },
});
