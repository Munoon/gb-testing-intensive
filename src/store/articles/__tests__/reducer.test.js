import reducer from "../reducer";
import { getArticlesFailure, getArticlesRequest, getArticlesSuccess, toggleFavourite } from "../actions";

describe('articles reducer tests', () => {
  it("matches snapshot after getArticlesRequest", () => {
    const result = reducer(undefined, getArticlesRequest());

    expect(result).toMatchSnapshot();
  });

  it('matches snapshot after getArticlesSuccess', () => {
    const articles = [{ title: 'test' }, { title: 'test2' }];
    const action = getArticlesSuccess(articles);
    const result = reducer(undefined, action);

    expect(result).toMatchSnapshot();
  });

  it('matches snapshot after getArticlesFailure', () => {
    const action = getArticlesFailure(new Error('failed to request articles'));
    const result = reducer(undefined, action);

    expect(result).toMatchSnapshot();
  });

  it('matches snapshot after toggle favourites', () => {
    const action = toggleFavourite('TEST_ID');
    const state = { articles: [{ title: 'test', id: 'TEST_ID' }] };
    const result = reducer(state, action);

    expect(result).toMatchSnapshot();
  });

  it('matches snapshot after toggle favourites, that not exist', () => {
    const action = toggleFavourite('TEST_ID');
    const result = reducer(undefined, action);

    expect(result).toMatchSnapshot();
  });
});
