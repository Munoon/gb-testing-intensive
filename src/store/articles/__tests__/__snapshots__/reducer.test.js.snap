// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articles reducer tests matches snapshot after getArticlesFailure 1`] = `
Object {
  "articles": Array [],
  "request": Object {
    "error": [Error: failed to request articles],
    "status": 3,
  },
}
`;

exports[`articles reducer tests matches snapshot after getArticlesRequest 1`] = `
Object {
  "articles": Array [],
  "request": Object {
    "error": null,
    "status": 1,
  },
}
`;

exports[`articles reducer tests matches snapshot after getArticlesSuccess 1`] = `
Object {
  "articles": Array [
    Object {
      "title": "test",
    },
    Object {
      "title": "test2",
    },
  ],
  "request": Object {
    "error": null,
    "status": 2,
  },
}
`;

exports[`articles reducer tests matches snapshot after toggle favourites 1`] = `
Object {
  "articles": Array [
    Object {
      "favourite": true,
      "id": "TEST_ID",
      "title": "test",
    },
  ],
}
`;

exports[`articles reducer tests matches snapshot after toggle favourites, that not exist 1`] = `
Object {
  "articles": Array [],
  "request": Object {
    "error": null,
    "status": 0,
  },
}
`;
